sdks/java/io/clickhouse/src/main/java/org/apache/beam/sdk/io/clickhouse/ClickHouseIO.java

```diff
@@ -0,0 +1,364 @@
+    private static final Logger LOG = LoggerFactory.getLogger(WriteFn.class);
+    private static final int QUEUE_SIZE = 1024;
+
+    private ClickHouseConnection connection;
+    private ClickHouseStatement statement;
+
+    private ExecutorService executor;
+    private BlockingQueue<Row> queue;
+    private AtomicBoolean bundleFinished;
+    private Future<?> insert;
+
+    // TODO: This should be the same as resolved so that Beam knows which fields
+    // are being accessed. Currently Beam only supports wildcard descriptors.
+    // Once BEAM-4457 is fixed, fix this.
+    @FieldAccess("filterFields")
+    final FieldAccessDescriptor fieldAccessDescriptor = FieldAccessDescriptor.withAllFields();
+
+    public abstract String jdbcUrl();
+
+    public abstract String table();
+
+    public abstract TableSchema schema();
+
+    public abstract Properties properties();
+
+    public static <T> WriteFn<T> create(
+        String jdbcUrl, String table, TableSchema schema, Properties properties) {
+      return new AutoValue_ClickHouseIO_WriteFn<>(jdbcUrl, table, schema, properties);
+    }
+
+    static String quoteIdentifier(String identifier) {
+      String backslash = "\\\\";
+      String quote = "\"";
+
+      return quote + identifier.replaceAll(quote, backslash + quote) + quote;
+    }
+
+    @VisibleForTesting
+    static String insertSql(TableSchema schema, String table) {
+      String columnsStr =
+          schema
+              .columns()
+              .stream()
+              .filter(x -> !x.materializedOrAlias())
+              .map(x -> quoteIdentifier(x.name()))
+              .collect(Collectors.joining(", "));
+      return "INSERT INTO " + quoteIdentifier(table) + " (" + columnsStr + ")";
+    }
+
+    @Setup
+    public void setup() throws SQLException {
+      connection = new ClickHouseDataSource(jdbcUrl(), properties()).getConnection();
+      executor =
+          Executors.newSingleThreadExecutor(
+              new ThreadFactoryBuilder().setNameFormat("clickhouse-jdbc-%d").build());
+      queue = new ArrayBlockingQueue<>(QUEUE_SIZE);
+      bundleFinished = new AtomicBoolean(false);
+    }
+
+    @StartBundle
+    public void startBundle() throws SQLException {
+      statement = connection.createStatement();
+      bundleFinished.set(false);
+
+      // When bundle starts, we open statement and stream data in bundle.
+      // When we finish bundle, we close statement.
+
+      // For streaming, we create a background thread for http client,
+      // and we feed it through BlockingQueue.
+
+      insert =
+          executor.submit(
+              () -> {
+                try {
+                  statement.sendRowBinaryStream(
+                      insertSql(schema(), table()),
+                      stream -> {
+                        while (true) {
+                          Row row;
+                          try {
+                            row = queue.poll(1, TimeUnit.SECONDS);
```

Comment:
Sorry, forgot to include context that other connectors do retries and an inner batching, because, as I understand, there is guarantee that data will fit the memory.

With this approach we stream data as it arrives, so we don't have to keep it in memory.